README.md
# HealthDataInterpretation
LLM to answer Personal Health Data Questions


history.txt

Prompt: Provide information on when each test was conducted and the frequency of these tests. Explain why each test was administered. Include info on the medical necessity and relevance of each test. Given, the patient's value and whether it is high, low, or within the range, include the potential impact of any deviations to the norm on the patient's health. Do not assume whether the patient's value is high or low; only use the information given
Context: Hemoglobin a1c or HbA1c is hemoglobin with glucose irreversibly bound. It is a long-term marker of glucose in excess. It is used in montioring for diabetes. A normal value for HbA1c is between 4.7 to 8.5%.The patient's value for Hemoglobin A1C is 4.9.The patient's value for Hemoglobin A1C is within the range.
Response: The Hemoglobin A1C test was conducted on [date] and is typically done every three months for patients with diabetes. This test measures the average blood sugar levels over the past two to three months. It is administered to monitor and manage diabetes by providing an indication of how well blood sugar levels have been controlled over time.

The medical necessity of the Hemoglobin A1C test lies in its ability to assess long-term glucose control. It helps healthcare providers determine if a patient's diabetes management plan is effective or if adjustments need to be made. By monitoring the Hemoglobin A1C levels, healthcare professionals can make informed decisions regarding medication dosage, lifestyle modifications, and overall treatment plans.

In the case of the patient, their Hemoglobin A1C value is 4.9, which falls within the normal range of 4.7 to 8.5%. This indicates that their blood sugar levels have been well-controlled over the past few months. Deviations from the normal range can have significant impacts on a patient's health. 

If the Hemoglobin A1C value is consistently high, it suggests poor blood sugar control, which can increase the risk of complications associated with diabetes. These complications may include cardiovascular disease, kidney damage, nerve damage, and eye problems. On the other hand, if the Hemoglobin A1C value is consistently low, it may indicate the risk of hypoglycemia (low blood sugar), which can lead to symptoms such as dizziness, confusion, and even loss of consciousness.

Therefore, maintaining a Hemoglobin A1C value within the normal range is crucial for managing diabetes and reducing the risk of long-term complications. Regular monitoring of this test allows healthcare providers to make necessary adjustments to the treatment plan and provide appropriate guidance to the patient for better diabetes management.

Prompt: Provide information on when each test was conducted and the frequency of these tests. Explain why each test was administered. Include info on the medical necessity and relevance of each test. Given, the patient's value and whether it is high, low, or within the range, include the potential impact of any deviations to the norm on the patient's health. Do not assume whether the patient's value is high or low; only use the information given
Context: A lipid panel measures certain the amount of different fat molecules in blood including cholesterol and tryglicerides. These are tested for buildup that can lead to cardiovascular problems if these values are too high. This is usually a routine test, but can be done to diagnose liver disease, pancreatitis, chronic kidney disease, and hypothyroidism.
Total cholesterol is a patient's overall cholesterol level. A normal range is between 150 to 199 mg/dL.
Low-density lipoprotein cholesterol or LDL is considered bad cholesterol and can collect in blood vessels and increase risk of cardiovascular disease. This should normally be below 130 mg/dL.
High-density lipoprotein cholesterol or HDL is considered good cholesterol and can help decrease the buildup of LDL in blood vessels. This should normally be greater than 40 mg/dL.
Tryglicerides are a type of fat gotten from eating food. Excess amounts of it are associated with cardiovascular disease and pancreatic inflammation. A normal value of this is under 150 mg/dL. A borderline range is between 150 and 199 mg/dL. A high range is between 200 and 499 mg/dL. A very high value is above 500 mg/dL.The patient's value for Total Cholesterol is 125.The patient's value for Total Cholesterol is low.The patient's value for Triglycerides is 90.The patient's value for Triglycerides is within the range.The patient's value for HDL Cholesterol is 50.The patient's value for HDL Cholesterol is within the range.The patient's value for HDL Cholesterol is 50.The patient's value for HDL Cholesterol is within the range.The patient's value for Non-HDL Cholesterol is 75.The patient's value for CHOL/HDL Ratio is 2.5.The patient's value for CHOL/HDL Ratio is within the range.
Response: The information provided does not include details about when each test was conducted or the frequency of these tests. However, I can provide general information about the medical necessity and relevance of each test.

A lipid panel is typically conducted as a routine test to assess a person's risk for cardiovascular disease. It measures the levels of different types of fats in the blood, including cholesterol and triglycerides. High levels of these fats can contribute to the development of plaque in the arteries, increasing the risk of heart disease and stroke.

Total cholesterol is a measure of the overall cholesterol level in the blood. Low levels of total cholesterol, such as the patient's value of 125 mg/dL, are generally considered to be within the normal range. However, extremely low levels may be associated with certain health conditions and may warrant further investigation.

Triglycerides are a type of fat that the body uses for energy. The patient's value of 90 mg/dL falls within the normal range. Elevated triglyceride levels, especially above 150 mg/dL, can be a risk factor for cardiovascular disease.


flask-boilerplate/newdemo.py
import openai
import numpy as np
from config import OPENAI_API_KEY
openai.api_key = OPENAI_API_KEY

test_type = [
    {
        "name": "Lipid Panel",
        "contextFile": "/lipid_panel_context.txt",
        "subTests": [
            {
                "name": "Total Cholesterol",
                "shortened": "tc",
                "rangeFlag": 1,
                "rangeHigh": 199,
                "rangeLow": 150 
            },
            {
                "name": "Triglycerides",
                "shortened": "triglycerides",
                "rangeFlag": 1,
                "rangeHigh": 150,
                "rangeLow": 0 
            },
            {
                "name": "HDL Cholesterol",
                "shortened": "hdl",
                "rangeFlag": 1,
                "rangeHigh": np.inf,
                "rangeLow": 40 
            },
            {
                "name": "LDL Cholesterol",
                "shortened": "ldl",
                "rangeFlag": 1,
                "rangeHigh": 130,
                "rangeLow": 0 
            },
            {
                "name": "Non-HDL Cholesterol",
                "shortened": "non-hdl",
                "rangeFlag": 0,
                "rangeHigh": "none",
                "rangeLow": "none"
            },
            {
                "name": "CHOL/HDL Ratio",
                "shortened": "CHOL/HDL Ratio",
                "rangeFlag": 1,
                "rangeHigh": 5,
                "rangeLow": 0 
            }
        ]        
    },
    {
        "name": "Hemoglobin A1C",
        "contextFile": "hemoglobin_a1c_context.txt",
        "subTests": [
            {
                "name": "Hemoglobin A1C",
                "shortened": "HbA1c",
                "rangeFlag": 1,
                "rangeHigh": 8.5,
                "rangeLow": 4.7 
            }
        ]
    }
]



prompt = "Provide information on when each test was conducted and the frequency of these tests. Explain why each test was administered. Include info on the medical necessity and relevance of each test. Given, the patient's value and whether it is high, low, or within the range, include the potential impact of any deviations to the norm on the patient's health. Do not assume whether the patient's value is high or low; only use the information given"
def demo(data):
    context_file_name = test_type[data["testId"]]["contextFile"]
    file=open("flask-boilerplate/"+context_file_name,"r")
    context = "".join(file.readlines())
    for x in data["results"]:
        name = x["name"]
        value = x["value"]
        rangeFlag = 0
        rangeHigh = 0
        rangeLow = 0
        for y in test_type[data["testId"]]["subTests"]:
            if (y["name"] == name):
                rangeFlag = y["rangeFlag"]
                rangeHigh = y["rangeHigh"]
                rangeLow = y["rangeLow"]
        context += "The patient's value for " + name +  " is " +str(value) + "."
        if(rangeFlag == 1):
            if(value > rangeHigh):
                context += "The patient's value for " + name + " is high."
            elif(value < rangeLow):
                context += "The patient's value for " + name + " is low."
            else:
                 context += "The patient's value for " + name + " is within the range."

    # Step 1: send the conversation and available functions to GPT  
    messages = [{"role": "system", "content": "You are ChartChat the helpful patient data bot."},
                {"role": "assistant", "content": context},
                {"role": "user", "content": prompt},
                {"role": "assistant", "content": "if explaining a definition, answer using 1 global metaphor"},
                {"role": "assistant", "content": "answer where a young adult can understand"},
                {"role": "assistant", "content": "answer in paragraph format and use less than 250 words."}]
    response = openai.ChatCompletion.create(
        model="gpt-3.5-turbo-16k-0613",
        messages=messages,
        temperature = 0.2
    )
    response_message = response["choices"][0]["message"]
    # response_message = demo()
    # print(response_message)
    with open('history.txt', 'a') as f:
        writeList = ['\nPrompt: ', prompt, "\nContext: ", context, '\n', "Response: ", response_message["content"], '\n']
        f.writelines(writeList)
    f.close()
    return response_message["content"]


flask-boilerplate/app.py
#----------------------------------------------------------------------------#
# Imports
#----------------------------------------------------------------------------#

from flask import Flask, render_template, request
# from flask.ext.sqlalchemy import SQLAlchemy
import logging
from logging import Formatter, FileHandler
import os
import newdemo

#----------------------------------------------------------------------------#
# App Config.
#----------------------------------------------------------------------------#

app = Flask(__name__)
app.config.from_object('config')
#db = SQLAlchemy(app)

# Automatically tear down SQLAlchemy.
'''
@app.teardown_request
def shutdown_session(exception=None):
    db_session.remove()
'''

# Login required decorator.
'''
def login_required(test):
    @wraps(test)
    def wrap(*args, **kwargs):
        if 'logged_in' in session:
            return test(*args, **kwargs)
        else:
            flash('You need to login first.')
            return redirect(url_for('login'))
    return wrap
'''
#----------------------------------------------------------------------------#
# Controllers.
#----------------------------------------------------------------------------#

patientInformation = {
    "age": 35,
    "gender": "Male"
}

tests = [
    {
        "test": "Lipid Panel",
        "date": "10/10/23",
        "testId": 0,
        "results": [
            {
            "name": "Total Cholesterol",
            "normal": "150 - 199 mg/dl",
            "value": 125
            },
            {
            "name": "Triglycerides",
            "normal": "0.00 - 150.00 mg/dL",
            "value": 90
            },
            {
            "name": "HDL Cholesterol",
            "normal": "above 40 mg/dL",
            "value": 50
            },
            {
            "name": "Non-HDL Cholesterol",
            "normal": "30 mg/dL above LDL",
            "value": 75
            },
            {
            "name": "CHOL/HDL Ratio",
            "normal": "0.0 - 5.0",
            "value": 2.5
            }
        ]
    },
    {
        "test": "Hemoglobin A1C",
        "date": "10/10/23",
        "testId": 1,
        "results": [
            {
            "name": "Hemoglobin A1C",
            "normal": "4.7 - 8.5 %",
            "value": 4.9
            }
        ]
    },
    {
        "test": "HSV TYPE 2 SPECIFIC ANTIBODY, IGG",
        "date": "10/10/23",
        "testId": 2,
        "results": [
            {
            "name": "Herpes Simplex Virus 2 IgG",
            "normal": "Negative",
            "value": "Negative"
            },
            {
            "name": "Herpes Simples Virus 2 IgG, Quant",
            "normal": "0.0 - 0.8 AI",
            "value": "<0.2"
            }
        ]
    },
]

def getInfo(data):
    #will be used to call chatGPT but for now it is just gonna have some words
    return newdemo.demo(data).split("\n")
   
@app.route('/')
def home():
    return render_template('pages/placeholder.home.html', tests=tests, len=len(tests))


@app.route('/about')
def about():
    return render_template('pages/placeholder.about.html')

@app.route('/testData/<int:test_index>')
def testData(test_index):
    data = tests[test_index]
    info = getInfo(data)
    return render_template('pages/placeholder.testData.html', data=data, info=info)

if not app.debug:
    file_handler = FileHandler('error.log')
    file_handler.setFormatter(
        Formatter('%(asctime)s %(levelname)s: %(message)s [in %(pathname)s:%(lineno)d]')
    )
    app.logger.setLevel(logging.INFO)
    file_handler.setLevel(logging.INFO)
    app.logger.addHandler(file_handler)
    app.logger.info('errors')

#----------------------------------------------------------------------------#
# Launch.
#----------------------------------------------------------------------------#

# Default port:
if __name__ == '__main__':
    app.run()

# Or specify port manually:
'''
if __name__ == '__main__':
    port = int(os.environ.get('PORT', 5000))
    app.run(host='0.0.0.0', port=port)
'''
